### ISSUE TITLE
Error in extendedP2proc

### ISSUE THREAD

**Issue by Dimmiso** [kharchenkolab/pagoda2#135]
Hi, 
Thanks for supporting conos and pagoda2! 😊
I aim to visualisation of conos object with pagoda2 app. In general it works well, but I experience problem with runnig extendedP2proc() to calculate GO pathway overdispersion. This procedure works for me for demo walkthrough, http://pklab.med.harvard.edu/pagoda2.walkthrough.html , but not for my data 
(https://drive.google.com/file/d/1Z2H0aHKOtb3hUHbLsn_DfmJ0FttO-5fg/view?usp=sharing)

`cdl.1 <- list('WG18008'=as(counts.WG18008,'dgCMatrix'),
              'WG17019'=as(counts.WG17019,'dgCMatrix'),
              'WG17020'=as(counts.WG17020,'dgCMatrix'))`

`p2l.1 <- lapply(cdl.1, basicP2proc, n.cores=30, min.cells.per.gene=0, n.odgenes=3e3, get.largevis=F, get.tsne=TRUE, make.geneknn=FALSE, nPcs=30)`

This list of  pagoda2 objects works well as an entry point for conos flow the whole way down to getting .bin file for pagoda2 app. However, if I try to use this list to calculate GO pathway overdispersion  

`ext.res <- extendedP2proc(p2l.1, organism = "hs")`

 I get:  Error in p2$testPathwayOverdispersion(setenv = go.env, verbose = TRUE,  : 
 attempt to apply non-function

It would be idal to work on this pagoda2 objects list, but I even tried with just one matrix.

`p2l.1.single <- basicP2proc(counts.WG17019, n.cores=30, min.cells.per.gene=0, n.odgenes=2e3, get.largevis=F, get.tsne=T, make.geneknn=F, nPcs=30) `

`ext.res <- extendedP2proc(p2l.1.single, organism = "hs")`

Based on messeges, the process moves somewhat further than in previous case, but I get error anyway:

determining valid pathways
processing 3379 valid pathways
scoring pathway od signifcance
Error in 1:ncol(pwpca[[i]]$xp$rotation) : argument of length 0
In addition: Warning message:
  In mclapply(..., mc.cores = n.cores, mc.preschedule = mc.preschedule) :
  scheduled cores 1, 2, 3, 4, 5, 6, 12, 15, 20, 21 did not deliver results, all values of the jobs will be affected

Is there any way to get GO pathways overdispersion data in my situation with can be later used for visualisation in pagoda2 app together with conos object? It would be nice to use pagoda2 object list, i.e. cdl.1 in my case, as an entry point. 

Another question. I found only this video tutorial for pagoda2, https://www.youtube.com/watch?v=xzpG1ZYE4Og
But it seems not having sound. Do you know if it is ment to be like this?

Thanks a lot! 
Dmitry

**Comment by evanbiederstedt** [comment]
Hi @Dimmiso 

RE: `basicP2proc()` and `extendedP2proc()`

I think the issue here is that I should improve the documentation. Based on user feedback, these functions cause more trouble than they're worth---I never use them myself. 

These functions essentially are wrapper functions which run all pagoda2 processing functions at once. 

In principle, this should work splendidly, as it's a single function which processes the data for users. In practice, this rarely works: the data may have issues, and the steps encompassed within `basic2proc()` do not give the user informative errors. Users need to make pre-processing & QC decisions based on their data, and these functions don't allow users to easily do that---the entire design of them is misguided tbh.

Please use these steps instead:

https://github.com/kharchenkolab/pagoda2/blob/main/doc/pagoda2.walkthrough.md#ii-in-depth-processing-and-analysis

**RE: pagoda2 app question**

> Another question. I found only this video tutorial for pagoda2, https://www.youtube.com/watch?v=xzpG1ZYE4Og
But it seems not having sound. Do you know if it is ment to be like this?

Yes, this is how the tutorial was created. 

I think based on user feedback, video tutorial aren't enough---users want informative documentation. If we do have video tutorials, we should also have them more informative with narration. 

The entire reason we haven't done this is that a newer application is in the works; we'd therefore prefer not to implement feature requests, etc. for pagoda2 anymore, as time could be invested elsewhere.

**Comment by Dimmiso** [comment]
Hi Evan, 

thanks a lot for detailed comments! I think I got the point. If it is ok, please, keep the issue open for some days. Following your tip I am trying to find my way from conos integration to pathway and hierarchical overdispersion analysis. In couple days I will test different pipelines and will be back. 
Thanks a lot for your assistance!

Best, Dmitry

**Comment by Dimmiso** [comment]
Hi Evan, 

I am not sure if I should keep the question below within current issue or it is better to initiate a new issue. Hi, 

I am still trying to nail reproducible errors in converting my conis objects into p2 app, but one problem seems to be quite consistant . GO environment which is produced within a pipeline described in the walkthrough you refer to, technically works for me: 

`ids <- unlist(lapply(mget(colnames(con.1$counts), org.Hs.egALIAS2EG, ifnotfound=NA), function(x) x[1]))
rids <- names(ids)
names(rids) <- ids
go.env.1 <- list2env(eapply(org.Hs.egGO2ALLEGS,function(x) as.character(na.omit(rids[x]))))`

However, in the output go.env.1 probably, there is a “systematic error”. As example: 
According to go.env.1 term “Thermoception” (View(go.env.1[['GO:0050955']]) contains genes EAA4,  VR1, VR1, ANKTM1, WDR47, LTRPC6. However correct genes for this term GO term are GRIK2, RHO, TRPV1,TRPA1,TRPM8. 

(I checked it, probably in a bit bulky way, but still: 

`GO2ALLEGS.mm.DF <- data.frame(org.Mm.egGO2ALLEGS)
list.thermo <- Term(GOTERM)[grepl("^thermoception", Term(GOTERM))]
GO.query.list <- names(list.thermo)
spec.gene.GO.query <- GO2ALLEGS.DF[GO2ALLEGS.DF$go_id %in% GO.query.list,]
spec.genes.IDs <- unique(spec.gene.GO.query$gene_id)
zzz <-egSYMBOL.DF[egSYMBOL.DF$gene_id %in% spec.genes.IDs,]
unique(zzz$symbol))`

Superficial check-up indicated that it is the case at least for some other GO terms as well. 
Can you, please, cross-check the flow above for generating go.env.1 and help to fix it if it indeed contains such problem?

Thanks a lot!

Dmitry

**Comment by evanbiederstedt** [comment]
Thanks @Dimmiso 

I'm happy to look at the above. If you could help me out a bit more to quickly do this though:

Could you give me the R code (from start to finish) +/- the data to reproduce this behavior? Based upon this, I could provide a fix. Without it, there's some guess work involved (and it takes longer to accomplish). 

Thanks, Evan

**Comment by Dimmiso** [comment]
Hi, 

Thanks a lot for being ready to help!  I am sorry for delay, our cluster was inaccessible for last two days.

My con.1 object is here in .rds form 
https://drive.google.com/file/d/1IOIBHz9wOg9g2JZt6Cy-kRRbkh1zzqRo/view?usp=sharing

The code below is from p2 walkthrough:

`library(GO.db)`
`library(org.Hs.eg.db)`
`ids <- unlist(lapply(mget(colnames(con.1$counts), org.Hs.egALIAS2EG, ifnotfound=NA), function(x) x[1]))`
`rids <- names(ids)`
`names(rids) <- ids`
`go.env.1 <- list2env(eapply(org.Hs.egGO2ALLEGS,function(x) as.character(na.omit(rids[x]))))`

Now if one checks what genes are included in term GO:0050955 (which is "Thermoception") according generated environment, 

`View(go.env.1[['GO:0050955']])`

it returns: EAA4,  VR1, VR1, ANKTM1, WDR47, LTRPC6. However, the correct list of genes for term 'GO:0050955' is: 
GRIK2, RHO, TRPV1,TRPA1,TRPM8

I used the following code to confirm it: 
`GO2ALLEGS.DF <- data.frame(org.Hs.egGO2ALLEGS)`
`egSYMBOL.DF <- data.frame(org.Hs.egSYMBOL)`

`list.thermo <- Term(GOTERM)[grepl("^thermoception", Term(GOTERM))]`
`GO.query.list <- names(list.thermo)`
`spec.gene.GO.query <- GO2ALLEGS.DF[GO2ALLEGS.DF$go_id %in% GO.query.list,]`
`spec.genes.IDs <- unique(spec.gene.GO.query$gene_id)`
`zzz <-egSYMBOL.DF[egSYMBOL.DF$gene_id %in% spec.genes.IDs,]`
`unique(zzz$symbol)`

GRIK2, RHO, TRPV1,TRPA1,TRPM8

Can you, please, check and address this issue if the mistake is not on my side?
 
Thanks a lot supporting conos and pagoda2! 

Dmitry

**Comment by evanbiederstedt** [comment]
Hi @Dimmiso 

I just had time to look at this. 

So, in the pagoda2 tutorial, we have the following:

```
# translate gene names to ids
ids <- unlist(lapply(mget(colnames(r$counts), org.Hs.egALIAS2EG, ifnotfound=NA), function(x) x[1]))
```

We access the `counts` field from the Pagoda2 object `r`, which is detailed here: https://github.com/kharchenkolab/pagoda2/blob/main/R/Pagoda2.R#L40


In the above code, `con.1$counts` would be NULL. 

```
> con.1$counts
NULL
```

This is because a Conos object does not have a `$counts` field. The user inputs multiple samples within `Conos`: https://github.com/kharchenkolab/conos/blob/main/R/conclass.R#L26-L27

Each of these samples should have a counts field. 

I think that's the source of the bug here. 

On the other topic, I'll update the walkthrough shortly.

Best, Evan
